IC3 ACS - Auth] <stable/2025-06-30> ACS Token Exchange for Teams Phone Extensibility

paveldostalms opened the pull request:
# Data Plane API Specification Update Pull Request

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->
**APIs supporting authorization mechanism and token exchange for Teams Phone Extensibility**
**New Identifier Model representing Teams Phone Extensibility user in ACS**
**Stable API containing part of changes approved for beta preview**
(preview PRs https://github.com/Azure/azure-rest-api-specs/pull/32703 https://github.com/Azure/azure-rest-api-specs/pull/32704)

### Teams Phone Extensibility
* Enhances Teams phone functionality via ACS SDKs to meet advanced voice needs in the Contact Center as a Service(CCaaS) space. ​
* Enables interoperability between Teams-owned phone numbers and ACS Call Automation, supporting:​
  * Secure handover of calls from Teams to ACS Call Automation​
  * Forwarding calls back to Teams from ACS Call Automation​

### Multi-persona
* Application of Entra ID for ACS in the Teams Phone Extensibility context​
* The data of a Teams user is separate from the data of the same Entra ID user when using a CCaaS application (call history, endpoint ringing, presence must be separate between the apps)​

### Why Authorization
* Azure RBAC only works for users in the same tenant as the ACS resource​
* Common scenario for CCaaS: an ISV provides CCaaS for somebody else; the ISV owns the ACS resource, but the Teams users are in different tenants ​
* Azure RBAC cannot be used​, no plans for Azure RBAC to become multi tenant​
* We were told to build our own authorization mechanism to setup authorization for users to the ACS resource​

### Why Token Exchange to Skype tokens
* There are current limitations that require a token exchange with the ACS token:​
  * IC3 services do not support Entra tokens and rely on Skype tokens.​
  * Entra tokens are not scoped to ACS resources​
* Our long-term aim is to have a trusted Entra token for all ACS and IC3 services.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## API Info: The Basics

Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [X] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: https://microsoft.sharepoint-df.com/:w:/t/IC3SDK/EVJaUNZNBANLjr2fh3zX-mABhg1c5_T4iefrktNyu2Kg4g?e=yxutpL
* Previous API Spec Doc:
  * Previous Stable 
    * https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Identity/stable/2023-10-01/CommunicationIdentity.json
    * https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Common/stable/2023-11-15/common.json
  * Approved preview
    * https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Identity/preview/2025-03-02-preview/CommunicationIdentity.json
    * https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Common/preview/2025-03-15-preview/common.json

* Updated paths:
  * new properties on PhoneNumberIdentifierModel, new TeamsExtensionUserIdentifierModel
  * new "/access/teamsExtension/:exchangeAccessToken" with post method
  * new "/access/teamsExtension/tenants/{tenantId}/assignments/{objectId}" with get, put, delete methods

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

https://apiview.dev/Assemblies/Review/0d3704e599a74da69d6b8a7eadb53d96?revisionId=7e5f925bafdc44dd88611c7db5872f8d

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

### Release planner

A [release plan](https://aka.ms/azsdkdocs/release-plans) should have been created. If not, please create one as it will help guide you through the REST API and SDK creation process. 

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories) 
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

paveldostalms committed:
Base commit

Base commit for common and identity

paveldostalms committed:
Updated Common swagger

Changes in Common swagger for TeamsExtension

paveldostalms committed:
Identity Swagger

Added TeamsExtension API to swagger

paveldostalms committed:
Prettier fixes

paveldostalms committed:
Fixed package path

paveldostalms committed:
Fixed api-version in sample

paveldostalms committed:
Removed unused example

paveldostalms committed:
Removing code for DurableIdentity

Forgotten changes from review that shouldn't got to this stable version

paveldostalms commented:
## Core Concepts

**Glossary**
*Teams Phone* - telephony service integrated with Microsoft Teams
*ACS* - Azure Communication Services

**Teams Phone extensibility** - the integration of Teams Phone and ACS, allowing the developers to leverage Teams Phone capabilities in their ACS based applications.
**Assignment of Teams Phone user** - allowing the Teams Phone users to use Teams Phone extensibility with defined ACS resource. Any single user can be assigned to any number of ACS resources. Only ACS resource owner can create or delete the assignment.
**Token exchange** - ACS doesn't accept the Entra access token of Teams Phone user directly. It is necessary to do the token exchange that takes the Teams Phone user token and returns the token usable for Teams Phone extension to work with specified ACS resource. 
*Note*: Calling of the token exchange would be wrapped in the SDK in most cases and the Teams Phone user doesn't need to care about it.

paveldostalms commented:
## Hero scenarios
Usage including context https://github.com/Azure/communication-preview/blob/master/Teams%20Phone%20Extensibility/teams-phone-extensibility-access-teams-phone.md


**As Communication Service resource owner, assigning Teams Phone user to ACS resource.**
For Teams Phone user with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.
Assignment created this way allows the Microsoft Teams user with oid _objectId_ from tenant _tenantId_ to use Entra application _appId_ to get an access token for using Teams Phone extensibility with ACS resource.

```
async Task CreateTeamsExtensionAssignmentExample(ObjectId objectId, ObjectId appId, TenantId tenantId, CancellationToken cancellationToken)
{
    var assignmentRequest = new TeamsExtensionAssignmentCreateOrUpdateRequest(TeamsPhonePrincipalType.User, new[] { appId });
    
    var json = JsonConvert.SerializeObject(assignmentRequest);
    var httpRequest = new HttpRequestMessage(HttpMethod.Put, $"access/teamsExtension/tenants/{tenantId}/assignments/{objectId}?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };
    await AddHmacAuthenticationAsync(httpRequest);
    
    var assignmentResponse = await SendAsync<TeamsExtensionAssignmentResponse>(httpRequest, cancellationToken);
    var assignment = assignmentResponse.Value;
    
    Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
    Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
    Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
    Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
}
```

**As Communication Service resource owner, getting existing assignment of the Teams Phone user to ACS resource.**
For Teams Phone user with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.

```
async Task GetTeamsExtensionAssignmentExample(TenantId tenantId, ObjectId objectId, CancellationToken cancellationToken)
{
    var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"access/teamsExtension/tenants/{tenantId}/assignments/{objectId}?api-version=2025-03-02-preview");
    await AddHmacAuthenticationAsync(httpRequest);

    var assignmentResponse = await SendAsync<TeamsExtensionAssignmentResponse>(httpRequest, cancellationToken);
    var assignment = assignmentResponse.Value;

    Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
    Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
    Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
    Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
}
```
**As Communication Service resource owner, deleting existing assignment of the Teams Phone user to ACS resource.**
For Teams Phone user with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.
With assignment deleted, Teams Phone user can't get token for using Teams Phone extensibility with ACS resource any more and existing tokens would be revoked.

```
async Task DeleteTeamsExtensionAssignmentExample(TenantId tenantId, ObjectId objectId, CancellationToken cancellationToken)
{
    var httpRequest = new HttpRequestMessage(HttpMethod.Delete, $"access/teamsExtension/tenants/{tenantId}/assignments/{objectId}?api-version=2025-03-02-preview");
    await AddHmacAuthenticationAsync(httpRequest);

    var response = await SendAsync<HttpResponseMessage>(httpRequest, cancellationToken);

    if (response.IsSuccessStatusCode)
    {
        Console.WriteLine("Assignment deleted successfully.");
    }
    else
    {
        Console.WriteLine($"Failed to delete assignment. Status code: {response.StatusCode}");
    }
}
```

**As Teams Phone user, use Entra User Access token to get access token for using Teams Phone extensibility with ACS resource**
ACS Resource is defined by request host.
This call won't be successful until ACS resource owner creates the assignment for the Teams Phone user in the ACS resource.

```
async Task ExchangeTeamsPhoneTokenExample(CancellationToken cancellationToken)
{
    var entraAccessToken = GetEntraAccessTokenAsync();
    var exchangeTokenRequest = new TeamsExtensionExchangeTokenRequest();

    var json = JsonConvert.SerializeObject(exchangeTokenRequest);
    var httpRequest = new HttpRequestMessage(HttpMethod.Post, $"access/teamsExtension/:exchangeAccessToken?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };

    httpRequest.Headers.Add(HeaderNames.Authorization, $"Bearer {entraAccessToken}");

    var response = await SendAsync<CommunicationIdentityAccessTokenResult>(httpRequest, cancellationToken);

    if (response.IsSuccessStatusCode)
    {
        Console.WriteLine($"Identity: {response.Value.Identity}");
        Console.WriteLine($"Access token: {response.Value.AccessToken}");
    }
    else
    {
        Console.WriteLine($"Failed to exchange token. Status code: {response.StatusCode}");
    }
}
```

weidongxu-microsoft commented on specification/communication/data-plane/Identity/stable/2025-06-30/CommunicationIdentity.json:
```diff
@@ -0,0 +1,695 @@
+        }
+      }
+    },
+    "TeamsExtensionExchangeTokenRequest": {
+      "type": "object",
+      "properties": {}
+    },
```
Is this schema exists here for future expansion? There is currently no property in it.

paveldostalms replied:
Yes, exactly. We want to prevent breaking changes in the case of future optional parameters. We expect future expansion of the scope of Teams Phone Extensibility feature and it may lead to the necessity to add some parameters to this call.

weidongxu-microsoft commented:
Confirm that 2025-06-30 is a subset of reviewed 2025-03-02-preview

Compare to prior stable, changes are

Common
```
6c6
<     "version": "2023-11-15"
---
>     "version": "2025-06-30"
81c81,82
<         "microsoftTeamsApp"
---
>         "microsoftTeamsApp",
>         "teamsExtensionUser"
114a116,119
>         },
>         "teamsExtensionUser": {
>           "$ref": "#/definitions/TeamsExtensionUserIdentifierModel",
>           "description": "The Microsoft Teams Extension user."
140c145,153
<           "description": "The phone number in E.164 format."
---
>           "description": "The phone number, usually in E.164 format."
>         },
>         "isAnonymous": {
>           "type": "boolean",
>           "description": "True if the phone number is anonymous. By default false if missing. If the phone number is anonymous, the value will be the string 'anonymous'."
>         },
>         "assertedId": {
>           "type": "string",
>           "description": "The asserted Id of the phone number. An asserted Id gets generated when the same phone number joins the same call more than once."
153c166
<           "description": "The Id of the Microsoft Teams user. If not anonymous, this is the AAD object Id of the user."
---
>           "description": "The Id of the Microsoft Teams user. If not anonymous, this is the Entra ID object Id of the user."
178a192,218
>         }
>       }
>     },
>     "TeamsExtensionUserIdentifierModel": {
>       "description": "A Microsoft Teams Phone user who is using a Communication Services resource to extend their Teams Phone set up.",
>       "type": "object",
>       "required": [
>         "userId",
>         "tenantId",
>         "resourceId"
>       ],
>       "properties": {
>         "userId": {
>           "type": "string",
>           "description": "The Id of the Microsoft Teams Extension user, i.e. the Entra ID object Id of the user."
>         },
>         "tenantId": {
>           "type": "string",
>           "description": "The tenant Id of the Microsoft Teams Extension user."
>         },
>         "resourceId": {
>           "type": "string",
>           "description": "The Communication Services resource Id."
>         },
>         "cloud": {
>           "$ref": "#/definitions/CommunicationCloudEnvironmentModel",
>           "description": "The cloud that the Microsoft Teams Extension user belongs to. By default 'public' if missing."
```

Identity
```
diff Identity/stable/2023-10-01/CommunicationIdentity.json Identity/stable/2025-06-30/CommunicationIdentity.json
6c6
<     "version": "2023-10-01"
---
>     "version": "2025-06-30"
39c39
<               "$ref": "../../../Common/stable/2022-07-13/common.json#/definitions/CommunicationErrorResponse"
---
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
82c82
<               "$ref": "../../../Common/stable/2022-07-13/common.json#/definitions/CommunicationErrorResponse"
---
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
122c122
<               "$ref": "../../../Common/stable/2022-07-13/common.json#/definitions/CommunicationErrorResponse"
---
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
141c141
<         "summary": "Exchange an Azure Active Directory (Azure AD) access token of a Teams user for a new Communication Identity access token with a matching expiration time.",
---
>         "summary": "Exchange an Entra ID access token of a Teams user for a new Communication Identity access token with a matching expiration time.",
167c167
<               "$ref": "../../../Common/stable/2022-07-13/common.json#/definitions/CommunicationErrorResponse"
---
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
222c222
<               "$ref": "../../../Common/stable/2022-07-13/common.json#/definitions/CommunicationErrorResponse"
---
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
237a238,446
>     },
>     "/access/teamsExtension/:exchangeAccessToken": {
>       "post": {
>         "tags": [
>           "TeamsExtension"
>         ],
>         "summary": "Exchanges a Teams Phone token for an ACS user access token.",
>         "operationId": "TeamsExtensionToken_Exchange",
>         "consumes": [
>           "application/json"
>         ],
>         "produces": [
>           "application/json"
>         ],
>         "parameters": [
>           {
>             "in": "body",
>             "name": "body",
>             "description": "Request payload for the token exchange.",
>             "required": true,
>             "schema": {
>               "$ref": "#/definitions/TeamsExtensionExchangeTokenRequest"
>             }
>           },
>           {
>             "$ref": "#/parameters/ApiVersionParameter"
>           }
>         ],
>         "responses": {
>           "default": {
>             "description": "Error",
>             "schema": {
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
>             }
>           },
>           "200": {
>             "description": "Success",
>             "schema": {
>               "$ref": "#/definitions/CommunicationIdentityAccessTokenResult"
>             }
>           }
>         },
>         "x-ms-examples": {
>           "Exchange a Teams Phone token": {
>             "$ref": "./examples/ExchangeTeamsPhoneToken.json"
>           }
>         }
>       }
>     },
>     "/access/teamsExtension/tenants/{tenantId}/assignments/{objectId}": {
>       "get": {
>         "tags": [
>           "TeamsExtension"
>         ],
>         "summary": "Get Teams Phone access assignment by object id.",
>         "operationId": "TeamsExtensionAssignment_Get",
>         "produces": [
>           "application/json"
>         ],
>         "parameters": [
>           {
>             "in": "path",
>             "name": "tenantId",
>             "description": "Tenant Id we want to get the assignment for.",
>             "required": true,
>             "type": "string"
>           },
>           {
>             "in": "path",
>             "name": "objectId",
>             "description": "Object Id we want to get the assignment for.",
>             "required": true,
>             "type": "string"
>           },
>           {
>             "$ref": "#/parameters/ApiVersionParameter"
>           }
>         ],
>         "responses": {
>           "default": {
>             "description": "Error",
>             "schema": {
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
>             }
>           },
>           "200": {
>             "description": "Success",
>             "schema": {
>               "$ref": "#/definitions/TeamsExtensionAssignmentResponse"
>             }
>           }
>         },
>         "x-ms-examples": {
>           "Get Assignment": {
>             "$ref": "./examples/GetTeamsExtensionAssignment.json"
>           }
>         }
>       },
>       "put": {
>         "tags": [
>           "TeamsExtension"
>         ],
>         "summary": "Creates or replaces a Teams Phone access assignment.",
>         "operationId": "TeamsExtensionAssignment_Upsert",
>         "consumes": [
>           "application/json"
>         ],
>         "produces": [
>           "application/json"
>         ],
>         "parameters": [
>           {
>             "in": "path",
>             "name": "tenantId",
>             "description": "Tenant Id we want to update the assignment for.",
>             "required": true,
>             "type": "string"
>           },
>           {
>             "in": "path",
>             "name": "objectId",
>             "description": "Object Id we want to update the assignment for.",
>             "required": true,
>             "type": "string"
>           },
>           {
>             "$ref": "#/parameters/ApiVersionParameter"
>           },
>           {
>             "in": "body",
>             "name": "body",
>             "description": "Values we want to set.",
>             "required": true,
>             "schema": {
>               "$ref": "#/definitions/TeamsExtensionAssignmentCreateOrUpdateRequest"
>             }
>           }
>         ],
>         "responses": {
>           "default": {
>             "description": "Error",
>             "schema": {
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
>             }
>           },
>           "200": {
>             "description": "Success",
>             "schema": {
>               "$ref": "#/definitions/TeamsExtensionAssignmentResponse"
>             }
>           },
>           "201": {
>             "description": "Created",
>             "schema": {
>               "$ref": "#/definitions/TeamsExtensionAssignmentResponse"
>             }
>           }
>         },
>         "x-ms-examples": {
>           "Create or Update Assignment": {
>             "$ref": "./examples/CreateOrReplaceTeamsExtensionAssignment.json"
>           }
>         }
>       },
>       "delete": {
>         "tags": [
>           "TeamsExtension"
>         ],
>         "summary": "Removes a Teams Phone access assignment.",
>         "operationId": "TeamsExtensionAssignment_Delete",
>         "produces": [
>           "application/json"
>         ],
>         "parameters": [
>           {
>             "in": "path",
>             "name": "tenantId",
>             "description": "Tenant Id we want to remove the assignment for.",
>             "required": true,
>             "type": "string"
>           },
>           {
>             "in": "path",
>             "name": "objectId",
>             "description": "Object Id we want to remove the assignment for.",
>             "required": true,
>             "type": "string"
>           },
>           {
>             "$ref": "#/parameters/ApiVersionParameter"
>           }
>         ],
>         "responses": {
>           "default": {
>             "description": "Error",
>             "schema": {
>               "$ref": "../../../Common/stable/2025-06-30/common.json#/definitions/CommunicationErrorResponse"
>             }
>           },
>           "204": {
>             "description": "No Content"
>           }
>         },
>         "x-ms-examples": {
>           "Delete Assignment": {
>             "$ref": "./examples/DeleteTeamsExtensionAssignment.json"
>           }
>         }
>       }
368a578
>       "description": "A request to create or update a Teams Phone assignment.",
377c587
<           "description": "Azure AD access token of a Teams User to acquire a new Communication Identity access token.",
---
>           "description": "Entra ID access token of a Teams User to acquire a new Communication Identity access token.",
381c591
<           "description": "Client ID of an Azure AD application to be verified against the appid claim in the Azure AD access token.",
---
>           "description": "Client ID of an Entra ID application to be verified against the appid claim in the Entra ID access token.",
385c595
<           "description": "Object ID of an Azure AD user (Teams User) to be verified against the oid claim in the Azure AD access token.",
---
>           "description": "Object ID of an Entra ID user (Teams User) to be verified against the oid claim in the Entra ID access token.",
388a599,656
>     },
>     "TeamsExtensionExchangeTokenRequest": {
>       "type": "object",
>       "properties": {}
>     },
>     "TeamsExtensionAssignmentCreateOrUpdateRequest": {
>       "required": [
>         "principalType"
>       ],
>       "type": "object",
>       "properties": {
>         "principalType": {
>           "$ref": "#/definitions/TeamsExtensionPrincipalType"
>         },
>         "clientIds": {
>           "type": "array",
>           "items": {
>             "type": "string"
>           }
>         }
>       }
>     },
>     "TeamsExtensionAssignmentResponse": {
>       "required": [
>         "objectId",
>         "principalType",
>         "tenantId"
>       ],
>       "type": "object",
>       "properties": {
>         "objectId": {
>           "type": "string"
>         },
>         "tenantId": {
>           "type": "string"
>         },
>         "principalType": {
>           "$ref": "#/definitions/TeamsExtensionPrincipalType"
>         },
>         "clientIds": {
>           "type": "array",
>           "items": {
>             "type": "string"
>           }
>         }
>       }
>     },
>     "TeamsExtensionPrincipalType": {
>       "description": "The type of principal the assignment is for.",
>       "enum": [
>         "resourceAccount",
>         "user"
>       ],
>       "type": "string",
>       "x-ms-enum": {
>         "name": "TeamsExtensionPrincipalType",
>         "modelAsString": true
>       }
392a661
>       "x-ms-parameter-location": "client",
418a688
>         "format": "url",
Common subdirectories: Identity/stable/2023-10-01/examples and Identity/stable/2025-06-30/examples
```

AI summary
```
Common.diff Changes (2023-11-15 → 2025-06-30):
1. API version update to 2025-06-30
2. Added new identity type "teamsExtensionUser" to the available identifiers
3. Enhanced PhoneNumberIdentifierModel with new properties:
   - Modified phone number description to be more flexible ("usually in E.164 format")
   - Added `isAnonymous` boolean field
   - Added `assertedId` field for handling multiple call joins
4. Updated terminology from "AAD" (Azure Active Directory) to "Entra ID"
5. Added new `TeamsExtensionUserIdentifierModel` definition for Teams Phone integration with required fields:
   - userId (Entra ID object Id)
   - tenantId
   - resourceId
   - Optional cloud environment setting

Identity.diff Changes (2023-10-01 → 2025-06-30):
1. API version update to 2025-06-30
2. Updated common.json references to use version 2025-06-30
3. Updated terminology from "Azure AD" to "Entra ID" across descriptions
4. Added significant new Teams Extension functionality:
   - New endpoint `/access/teamsExtension/:exchangeAccessToken` for exchanging Teams Phone tokens
   - New endpoints under `/access/teamsExtension/tenants/{tenantId}/assignments/{objectId}` for managing Teams Phone access assignments:
     - GET: Retrieve assignments
     - PUT: Create/update assignments
     - DELETE: Remove assignments
5. Added new definitions:
   - TeamsExtensionExchangeTokenRequest
   - TeamsExtensionAssignmentCreateOrUpdateRequest
   - TeamsExtensionAssignmentResponse
   - TeamsExtensionPrincipalType (with values: resourceAccount, user)
```

paveldostalms commented:
## Analysis of changes against approved previews

**API changes in this PR** 
https://github.com/Azure/azure-rest-api-specs/compare/ac7eb63..acae393d

**API changes in previews** 
https://github.com/Azure/azure-rest-api-specs/compare/12d70cb..9db51de
https://github.com/Azure/azure-rest-api-specs/compare/a21763c..83dd09e

Copilot analysis of changes:
---

**Conversation Summary: Preview vs GA Comparison for ACS Token Exchange for Teams Phone Extensibility**

**1. Analysis of GA (2025-06) Version Diff**
- Introduced Teams Phone Extension user support and models
- GA version “2025-06-30” for Common and Identity specs
- Added new endpoints for Teams Extension assignment management and token exchange
- Enriched models with properties like `isAnonymous`, `assertedId`
- Adopted “Entra ID” terminology
- Updated error schema references

**2. Analysis of Preview (2025-03) Version Diff**
- Introduced preview versions “2025-03-15-preview” and “2025-03-02-preview”
- Added Teams Phone Extension models and endpoints (matching GA)
- Added Entra assignment management endpoints (bulk/list/single assignment PATCH/GET/PUT/DELETE)
- Enhanced identity creation with `customId` and retrieval by custom ID (`GET /identities/{id}`)
- Added `customId` and `lastTokenIssuedAt` to identity model
- Many new/updated example files for preview endpoints and models

**3. Comparison: What’s in Preview but NOT in GA**
- Entra assignment management endpoints/models (bulk and per-object assignment)
- Identity management by customId (POST/GET /identities with customId)
- `lastTokenIssuedAt` property on identities
- Example files for Entra assignment flows

**4. Comparison: What’s in GA but NOT in Preview**
- Only GA versions (2025-06-30) of endpoints/models
- Stable error schema references

**5. Features in Both**
- Teams Phone Extension support (models/endpoints)
- `isAnonymous` and `assertedId` on phone number models
- “Entra ID” terminology
- Example files for Teams Extension scenarios

**Summary Table:**
| Feature                                 | Preview | GA      |
|------------------------------------------|---------|---------|
| Entra assignment endpoints/models        | Yes     | No      |
| Teams Extension assignment endpoints     | Yes     | Yes     |
| customId for identity                    | Yes     | No      |
| GET /identities/{id}                     | Yes     | No      |
| lastTokenIssuedAt property               | Yes     | No      |
| Stable error response schemas            | No      | Yes     |
| Versioned as 2025-06-30                  | No      | Yes     |
| isAnonymous/assertedId for phone number  | Yes     | Yes     |

_If you need a more granular endpoint-by-endpoint or model-by-model breakdown, let me know!_

---

Let me know if you'd like to further customize this summary before posting!

mikekistler reviewed (approved):
Looks good! 👍

paveldostalms committed:
Merge branch 'main' into paveldostal/communication/teamsextension_ga
